### PR TITLE
yunfaavatar: init at 0.2.0

### PR DESCRIPTION
**Pull request by yunfachi**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by cafkafk — CHANGES_REQUESTED** [review]
Hey, sorry to see you've been left hanging for so long. 

---

My first comment here is this fails to conform to [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

You should change it so that your first commit in this PR is just `maintainers: added <your handle>`, and then all the rest of your changes should be in a single commit `yunfaavatar: init at 0.1.0`.

To read more about adding yourself to the maintainer file, you can look here https://github.com/NixOS/nixpkgs/tree/master/maintainers.

---

If you have any questions about anything, please feel free to ask!

**Review by cafkafk — CHANGES_REQUESTED** [review]
Hey, so it seems your current commits messages don't conform, no worries! For comparison, here's how this should look.



You can also go check out that PR to see how changes are structured #256675. What you want to change this is to use `git rebase -i HEAD~4` (here 4 it's number of commits you've made + 1), and then it will show you some text explaining it to you. DW, this is kinda hard.

If you can't get it to work, you can try a git tool, I don't know which ones are good.

Also, if you need help with doing this, or either approach I mentioned doesn't work out for you, please feel free to ask.

File: pkgs/applications/misc/yunfaavatar/default.nix
Comment (eclairevoyant) [inline pkgs/applications/misc/yunfaavatar/default.nix:14]:
```suggestion
  version = "0.1.0";
```
release versions preferred when possible

File: pkgs/applications/misc/yunfaavatar/default.nix
Comment (eclairevoyant) [inline pkgs/applications/misc/yunfaavatar/default.nix:19]:
```suggestion
    rev = version;
```

File: pkgs/applications/misc/yunfaavatar/default.nix
Comment (eclairevoyant) [inline pkgs/applications/misc/yunfaavatar/default.nix:20]:
```suggestion
    hash = "sha256-JbolO1I4K9+RBFJ/U611En4Rx5cG2+z1IcxYEeiaq6Q=";
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

**Review by eclairevoyant — APPROVED** [review]

**Review by cafkafk — CHANGES_REQUESTED** [review]
This doesn't seem to be in `pkgs/top-level/all-packages` and so people can't actually use it. You could add it there, but another approach that was recently introduce is moving it to `pkgs/by-name`.

`mkdir -p pkgs/by-name/yu/yunfaavatar; mv pkgs/applications/misc/yunfaavatar/default.nix pkgs/by-name/yu/yunfaavatar/package.nix`

When you've done that, you might wanna `git commit --amend` your changes instead of making a new commit.

Good luck!

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (cafkafk) [inline pkgs/by-name/yu/yunfaavatar/package.nix:26]:
```suggestion
  buildInputs = with pkgs; [ bash imagemagick ];
```

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (cafkafk) [inline pkgs/by-name/yu/yunfaavatar/package.nix:4]:
```suggestion
```

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (cafkafk) [inline pkgs/by-name/yu/yunfaavatar/package.nix:35]: Descriptions can't end in a `.`, I know this is an edge case, but to make your life easier you could just rewrite it like this e.g.:
```suggestion
    description = "Utility for automatic centralized changing of avatar in Github, Discord, Steam, Hypixel, and many more";
```

**Review by cafkafk — CHANGES_REQUESTED** [review]
Probably final change request before approval, nice work so far!

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (eclairevoyant) [reply pkgs/by-name/yu/yunfaavatar/package.nix]: I missed this entirely. These should be inputs.

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (eclairevoyant) [inline pkgs/by-name/yu/yunfaavatar/package.nix:9]: Please don't use `pkgs` here. Individual inputs should be listed out (bash and imagemagick) to allow overriding

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (eclairevoyant) [inline pkgs/by-name/yu/yunfaavatar/package.nix:25]:
```suggestion
  buildInputs = [ bash imagemagick ];
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (K900) [inline pkgs/by-name/yu/yunfaavatar/package.nix:5]: Unused.

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (K900) [inline pkgs/by-name/yu/yunfaavatar/package.nix:8]: Actually these are all unused.

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (K900) [inline pkgs/by-name/yu/yunfaavatar/package.nix:27]: What are you expecting this to do?

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (K900) [inline pkgs/by-name/yu/yunfaavatar/package.nix:23]: Probably not necessary.

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (K900) [inline pkgs/by-name/yu/yunfaavatar/package.nix:31]: Unused.

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (K900) [inline pkgs/by-name/yu/yunfaavatar/package.nix:40]: This doesn't need to be set if it matches the package name.

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (K900) [inline pkgs/by-name/yu/yunfaavatar/package.nix:37]: Can you maybe clarify the license situation here?

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (yunfachi) [reply pkgs/by-name/yu/yunfaavatar/package.nix]: which ones exactly?

**Review by K900 — CHANGES_REQUESTED** [review]

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (K900) [reply pkgs/by-name/yu/yunfaavatar/package.nix]: That's a question for you to answer, not me.

**Comment by K900** [comment]
I see you used the [`neofetch` expression](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/tools/misc/neofetch/default.nix) as a base for this. It's fine to use other expressions as a reference, but you need to understand the changes you're making here. You have unused arguments, `makeFlags` that don't match your own upstream makefile, a `makeWrapper` input that you then never use (and you should!), and a bunch of other nitpicks. Please figure out what actually needs to be done to make _your_ software build and run.

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (eclairevoyant) [reply pkgs/by-name/yu/yunfaavatar/package.nix:40]: that's not true.

Comment (K900) [reply pkgs/by-name/yu/yunfaavatar/package.nix:36]: My bad.

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (cafkafk) [inline pkgs/by-name/yu/yunfaavatar/package.nix:29]:
```suggestion
```

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (cafkafk) [inline pkgs/by-name/yu/yunfaavatar/package.nix:20]:
```suggestion
```

**Review by cafkafk — CHANGES_REQUESTED** [review]
You should probably remove these comments.

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (cafkafk) [inline pkgs/by-name/yu/yunfaavatar/package.nix:5]: Correct me if I'm wrong but this doesn't seem used.
```suggestion
```

Comment (K900) [reply pkgs/by-name/yu/yunfaavatar/package.nix]: It should probably be used to inject the dependencies.

**Review by cafkafk — APPROVED** [review]
LGTM ([if K900 thinks the makeWrapper thing is okay](https://github.com/NixOS/nixpkgs/pull/255688#discussion_r1342145365))

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (cafkafk) [reply pkgs/by-name/yu/yunfaavatar/package.nix]: ohh, didn't see you here you were hidden. It does do *things* on my machine, not sure it that's enough to verify it's actually working :woman_shrugging: 

```
nix run ./#yunfaavatar
[INFO] Loaded the config (/home/ces/.config/yunfaAvatar/config.conf)

[INFO] Downloading an avatar
[INFO] Avatar size is 41,889 bytes
[INFO] Converted avatar to base64

[INFO] Finished
```

Comment (cafkafk) [reply pkgs/by-name/yu/yunfaavatar/package.nix]: (also I can't unresolve)

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (K900) [inline pkgs/by-name/yu/yunfaavatar/package.nix:20]: These are used at runtime, correct? If so, you need to use `makeWrapper` to add them to PATH for the actual executable.

Comment (cafkafk) [reply pkgs/by-name/yu/yunfaavatar/package.nix:20]: `imagemagick` here, https://github.com/yunfachi/yunfaAvatar/blob/be8d383b6eba41a37e52f0e3128ed3a234d4875f/utils/avatar.sh#L31

(also the whole things a bash script)

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (K900) [inline pkgs/by-name/yu/yunfaavatar/package.nix:21]: These still aren't necessary, bash is part of stdenv, and you're referring to imagemagick in your `postInstall` directly.

Comment (yunfachi) [reply pkgs/by-name/yu/yunfaavatar/package.nix:21]: so i need to delete lines 20 and 21?

Comment (K900) [reply pkgs/by-name/yu/yunfaavatar/package.nix:21]: You can keep `strictDeps` (though whether it's useful given you only have one dependency is debatable), you don't need the `buildInputs`.

Comment (yunfachi) [reply pkgs/by-name/yu/yunfaavatar/package.nix]: I removed `buildInputs` and did `postInstall`, can you please review?

File: pkgs/by-name/yu/yunfaavatar/package.nix
Comment (K900) [inline pkgs/by-name/yu/yunfaavatar/package.nix:28]: It should really go into $out/bin, otherwise it won't be added to PATH.

**Comment by K900** [comment]
This isn't going to work when all of your stuff is in `$out/opt`. Why `opt`, anyway?

**Comment by yunfachi** [comment]
> This isn't going to work when all of your stuff is in `$out/opt`. Why `opt`, anyway?

`opt` contains folders with bash scripts that do not belong in `bin`, but are imported in the main script. in `bin` I create a symbolic link of the main script from `opt`

**Comment by K900** [comment]
You might want to put that stuff into `$out/lib` then, and please make sure the package actually works this time - or, even better, add a simple test so it keeps being checked in the future.

**Comment by yunfachi** [comment]
in my case `wrapProgram $out/bin` doesn't work because `$out/bin/yunfaavatar` is a symbolic of `$out/lib/yunfaavatar`. everything works for me now

**Comment by yunfachi** [comment]
> You might want to put that stuff into `$out/lib` then, and please make sure the package actually works this time - or, even better, add a simple test so it keeps being checked in the future.

I did as you said, but `wrapProgram $out/bin` doesn't work because `$out/bin/yunfaavatar` is a symbolic of `$out/lib/yunfaavatar`. everything works for me now

**Comment by K900** [comment]
Now you have the binary in $out/lib, which means it won't be on PATH by default

**Comment by yunfachi** [comment]
> Now you have the binary in $out/lib, which means it won't be on PATH by default

what is the difference between a symbolic in bin that leads to the main script in lib (current) and the main script in bin? I would do it the second way, but I can’t/don’t fully understand how to import scripts from lib while in bin

**Comment by K900** [comment]
The way I'd do this personally is by putting the main script in $out/bin, and then using `$0` to determine the executable's path and then using the relative paths to source your additional library bits.

**Comment by K900** [comment]
Packaging LGTM now, with a few more nits on the overall process:

- please don't move tags around - you've retagged 0.2.0 on your upstream repo when incorporating these changes, which is fine for now, but will cause issues if you move tags after they're included in nixpkgs proper.
- a test would be nice to have, even if it's just something that verifies that `yunfaavatar --help` runs
- your application doesn't actually implement `--help` (or really any args handling, as far as I could see)

**Review by K900 — APPROVED** [review]

**Comment by K900** [comment]
I will not hold this change back though - those things can be added later. Thanks!

**Comment by yunfachi** [comment]
I'll add `—help` in the next update. Thank you

**Pull request merged**
